src/main/java/king/TaskList.java

```diff
@@ -0,0 +1,109 @@
+package king;
+
+import king.task.Task;
+
+import java.util.ArrayList;
+
+/**
+ * The TaskList class encapsulates a list of tasks.
+ * It provides operations to add, remove, retrieve, and manipulate tasks in the list.
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
+
+    /**
+     * Creates an empty TaskList.
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
+
+    /**
+     * Creates a TaskList with the specified tasks.
+     *
+     * @param tasks the list of tasks to be managed by this TaskList
+     */
+    public TaskList(ArrayList<Task> tasks) {
+        this.tasks = tasks;
+    }
+
+    /**
+     * Adds a task to the task list.
+     *
+     * @param task the task to be added
+     */
+    public void add(Task task) {
+        tasks.add(task);
+    }
+
+    /**
+     * Removes a task from the task list based on its index.
+     *
+     * @param index the index of the task to be removed
+     */
+    public void remove(int index) {
+        if (index >= 0 && index < tasks.size()) {
+            tasks.remove(index);
+        }
+    }
+
+    /**
+     * Retrieves a task from the task list based on its index.
+     *
+     * @param index the index of the task to be retrieved
+     * @return the task at the specified index, or null if the index is out of bounds
+     */
+    public Task getTask(int index) {
+        if (index >= 0 && index < tasks.size()) {
+            return tasks.get(index);
+        }
+        return null;
+    }
+
+    /**
+     * Returns the list of tasks managed by this TaskList.
+     *
+     * @return the list of tasks
+     */
+    public ArrayList<Task> getTaskList() {
+        return this.tasks;
+    }
+
+    /**
+     * Returns the number of tasks in the task list.
+     *
+     * @return the number of tasks
+     */
+    public int size() {
+        return tasks.size();
+    }
+
+    /**
+     * Marks a task as done based on its task number.
+     *
+     * @param taskNumber the task number to be marked as done
+     * @throws KingException if the task number is invalid
+     */
+    public void markTaskAsDone(int taskNumber) throws KingException {
```

Comment:
The exception name "KingException" should be replaced with a more standard and descriptive name, such as "InvalidTaskNumberException".